crates/kubelet/src/device_plugin_manager/pod_devices.rs

```diff
@@ -0,0 +1,157 @@
+pub type ContainerDevices = HashMap<String, ResourceAllocateInfo>;
+
+/// PodDevices contains the map of Pods to allocated devices
+#[derive(Clone)]
+pub struct PodDevices {
+    /// Map of devices allocated to the Pod keyed by Pod UID
+    allocated_devices: Arc<Mutex<HashMap<String, ContainerDevices>>>,
+    client: kube::Client,
+}
+
+impl PodDevices {
+    pub fn new(client: kube::Client) -> Self {
+        PodDevices {
+            allocated_devices: Arc::new(Mutex::new(HashMap::new())),
+            client,
+        }
+    }
+
+    /// get_active_pods activePods is a method for listing active pods on the node so the
+    /// amount of device plugin resources requested by existing pods could be counted
+    /// when updating allocated devices
+    pub async fn get_active_pods(&self) -> anyhow::Result<HashSet<String>> {
+        // TODO: should this be namespaced?
+        let pod_client: Api<Pod> = Api::all(self.client.clone());
+        let pods = pod_client.list(&ListParams::default()).await?;
+        Ok(pods
+            .iter()
+            .map(|pod| {
+                pod.metadata
+                    .uid
+                    .clone()
+                    .expect("Pod uid should always be set but was not")
+            })
+            .collect::<HashSet<String>>())
+    }
+
+    /// get_pods returns the UIDs of all the Pods in the `PodDevices` map
+    pub fn get_pods(&self) -> HashSet<String> {
+        self.allocated_devices
+            .lock()
+            .unwrap()
+            .iter()
+            .map(|(p_uid, _container_map)| p_uid.clone())
+            .collect::<HashSet<String>>()
+    }
+
+    pub fn remove_pods(&self, pods_to_remove: HashSet<String>) -> anyhow::Result<()> {
+        let mut allocated_devices = self.allocated_devices.lock().unwrap();
+        pods_to_remove.iter().for_each(|p_uid| {
+            allocated_devices.remove(p_uid);
+        });
+        Ok(())
+    }
+
+    /// Returns the device Ids of all devices requested by a container
+    pub fn get_container_devices(
+        &self,
+        resource_name: &str,
+        pod_uid: &str,
+        container_name: &str,
+    ) -> Option<HashSet<String>> {
+        let allocated_devices = self.allocated_devices.lock().unwrap().clone();
+        if let Some(container_devices) = allocated_devices.get(pod_uid) {
+            if let Some(resource_devices) = container_devices.get(container_name) {
+                if let Some(device_allocate_info) = resource_devices.get(resource_name) {
+                    return Some(device_allocate_info.device_ids.clone());
+                }
+            }
+        }
+        None
+    }
+
+    pub fn get_allocated_devices(&self) -> DeviceIdMap {
+        let allocated_devices = self.allocated_devices.lock().unwrap().clone();
+        let mut res: DeviceIdMap = HashMap::new();
+        allocated_devices
+            .iter()
+            .for_each(|(_p_uid, container_map)| {
```

Comment:
This is more pontification than actual feedback, but there is _so much_ nesting here. I wonder if there is another way we could model the data. I'll think on it some more and see if I come up with any fresh ideas

Reply:
I agree. I'd love if you have any thoughts. I was modeling after Kubernetes' [`PodDevices` struct](https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/kubelet/cm/devicemanager/pod_devices.go#L39), but I agree that it is incredibly nested. The goal is for `PodDevices` to contain the mapping of allocated resources to Pods.

Reply:
I added a comment that explains why it is so nested and that an in mem db may be a better solution in the future: 716fa458cb8bb8cf12